### PR TITLE
sorbet: Autogenerate the RBI file for `Homebrew::EnvConfig`

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- I'm not as happy with this as I was with my work on `utils/tty.rb`, it feels a lot hackier and prone to breaking - especially the `T.nilable` "detection".
- However, it generates an RBI file that passes `brew typecheck`, so we can clean up some `hidden-definitions` entries, getting further towards fully typed.

### CODE REVIEW

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# frozen_string_literal: true
+
+require_relative "../../global"
+require_relative "../../env_config"
+
+def nilable?(method)
+  %w[browser editor github_api_token].include?(method)
+end
+
+File.open("#{File.dirname(__FILE__)}/../../env_config.rbi", "w") do |file|
```
Comment (reitermarkus) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:11]:
```suggestion
File.open("#{__dir__}/../../env_config.rbi", "w") do |file|
```

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    else
+      nilable?(method) ? T.nilable(String) : String
+    end
+
+    file.write(<<-RUBY)
+  sig { returns(#{return_type}) }
+  def self.#{method}; end
+    RUBY
+
+    file.write("\n") unless methods.last == method
```
Comment (reitermarkus) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:42]:
```suggestion
    file.write("\n") unless method == methods.last
```

makes more sense to my brain. 😄

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  file.write("end\n")
```
Comment (reitermarkus) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:45]:
```suggestion
  file.puts "end"
```

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  dynamic_methods = {}
+  Homebrew::EnvConfig::ENVS.each do |env, hash|
+    next if Homebrew::EnvConfig::CUSTOM_IMPLEMENTATIONS.include?(env.to_s)
+
+    name = Homebrew::EnvConfig.env_method_name(env, hash)
+    dynamic_methods[name] = { default: hash[:default] }
+  end
+
+  methods = Homebrew::EnvConfig.methods(false).map(&:to_s).sort.select { |method| dynamic_methods.key?(method) }
```
Comment (reitermarkus) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:26]:
```suggestion
  methods = Homebrew::EnvConfig.methods(false).map(&:to_s).select { |method| dynamic_methods.key?(method) }.sort
```

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
+# typed: true
+# frozen_string_literal: true
+
+require_relative "../../global"
+require_relative "../../env_config"
+
+def nilable?(method)
```
Comment (reitermarkus) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:7]: This should be all without a `default` value, right? So I don't think we need this function.

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  methods = Homebrew::EnvConfig.methods(false).map(&:to_s).sort.select { |method| dynamic_methods.key?(method) }
+
+  methods.each do |method|
+    return_type = if method.end_with?("?")
+      T::Boolean
+    elsif dynamic_methods[method][:default].instance_of?(Integer)
+      Integer
+    else
+      nilable?(method) ? T.nilable(String) : String
+    end
```
Comment (reitermarkus) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:35]:
```suggestion
  methods.each do |method|
    return_type = if method.end_with?("?")
      T::Boolean
    elsif default = dynamic_methods[method][:default]
      default.class
    else
      T.nilable(String)
    end
```

File: Library/Homebrew/sorbet/custom_generators/env_config.rb (hunk 5 above)
Comment (issyl0) [reply Library/Homebrew/sorbet/custom_generators/env_config.rb:7]: You're correct! Thanks.

File: Library/Homebrew/sorbet/custom_generators/env_config.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    name = Homebrew::EnvConfig.env_method_name(env, hash)
+    dynamic_methods[name] = { default: hash[:default] }
+  end
+
+  methods = Homebrew::EnvConfig.methods(false).map(&:to_s).select { |method| dynamic_methods.key?(method) }.sort
+
+  methods.each do |method|
+    return_type = if method.end_with?("?")
+      T::Boolean
+    elsif (default = dynamic_methods[method][:default])F
```
Comment (issyl0) [inline Library/Homebrew/sorbet/custom_generators/env_config.rb:27]:
```suggestion
    elsif (default = dynamic_methods[method][:default])
```

**Comment by issyl0** [comment]
Thanks for the review and the fixes, @reitermarkus! ✨

File: Library/Homebrew/sorbet/custom_generators/env_config.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/sorbet/custom_generators/env_config.rb]: F as in fix. 😄

File: Library/Homebrew/resource.rb [hunk 8]
```diff
@@ -296,7 +296,7 @@ def determine_url_mirrors
 
     # PyPI packages: PEP 503 – Simple Repository API <https://peps.python.org/pep-0503>
     if Homebrew::EnvConfig.pip_index_url.present?
-      pip_index_base_url = Homebrew::EnvConfig.pip_index_url.chomp("/").chomp("/simple")
+      pip_index_base_url = T.must(Homebrew::EnvConfig.pip_index_url).chomp("/").chomp("/simple")
```
Comment (reitermarkus) [inline Library/Homebrew/resource.rb:299]:
```suggestion
  if (pip_index_url = Homebrew::EnvConfig.pip_index_url.presence)
      pip_index_base_url = pip_index_url.chomp("/").chomp("/simple")
```

**Review by reitermarkus — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me, thanks again @issyl0!

**Pull request merged**